src/bos/operators/power_operator_base.py

```diff
@@ -65,69 +65,101 @@ def _act(self, components) -> List[dict]:
+        If None, operate on all components in the map.
 
         Returns:
-          An index indicating the matched components location in the list
-          It returns None if there is no match.
-          :rtype: int
+          None
         """
-        for component in components:
-            if component_id == component['id']:
-                return components.index(component)
-        return None
+        if component_ids is None:
+            component_ids = list(component_id_map)
+        num_components = len(component_ids)
+        LOGGER.debug("_do_power_components called on %d components", num_components)
+        errors = self._my_power(component_ids)
+        if errors.error_code == 0:
+            return
+
+        if errors.nodes_in_error:
+            # Update any nodes with errors they encountered
+            for component_id, node_error in errors.nodes_in_error.items():
+                if component_id not in component_ids:
+                    LOGGER.debug("CAPMC error for xname that wasn't in request: %s error code=%s "
+                                 "message=%s", node_error.error_message, component_id)
+                    continue
+                LOGGER.debug("Component %s error code=%s message=%s", component_id,
+                             node_error.error_code, node_error.error_message)
+                component = component_id_map[component_id]
+                component['error'] = node_error.error_message
+                component['enabled'] = disable_based_on_error_xname_on_off(node_error.error_message)
+            return
+
+        # Errors could not be associated with a specific node.
+        if num_components == 1:
+            # Well, okay, I guess the error can be associated with this node
+            component_id = component_ids[0]
+            LOGGER.debug("Component %s error code=%s message=%s", component_id,
+                         errors.error_code, errors.error_message)
+            component = component_id_map[component_id]
+            component['error'] = errors.error_message
+            component['enabled'] = False
+            return
+
+        if num_components < 16:
+            # Ask CAPMC to act on them one at a time to identify
+            # nodes associated with errors.
+            for component_id in component_ids:
+                LOGGER.debug("Acting on component %s", component_id)
+                errors = self._my_power([component_id])
+                if errors.error_code == 0:
+                    continue
+                LOGGER.debug("Component %s error code=%s message=%s", component_id,
+                             errors.error_code, errors.error_message)
+                component = component_id_map[component_id]
+                component['error'] = errors.error_message
+                component['enabled'] = False
+            return
+
+        # Subdivide the nodes into 8 groups and recursively act on those
+        chunk_size = len(component_ids) // 8
```

Comment:
Wow... what is going on with capmc to require this monstrocity?

Reply:
It's just that sometimes CAPMC returns very vague failures, like "errors encountered with 1/2700 xnames". And currently BOS v2 responds to this by retrying each one individually, which takes a very long time when it is that many nodes. :)

Reply:
This change will not help if you are in a situation where more than around 10% of the nodes are failing in this way, but if you're in that situation, you've probably got bigger problems anyway.